src/mcts/search.cc

```diff
@@ -1145,8 +1145,49 @@ SearchWorker::NodeToProcess SearchWorker::PickNodeToExtend(
       }
       return NodeToProcess::Collision(node, depth, collision_limit);
     }
-    // Either terminal or unexamined leaf node -- the end of this playout.
-    if (node->IsTerminal() || !node->HasChildren()) {
+    // If terminal, we either found a twofold draw to be reverted, or
+    // reached the end of this playout.
+    if (node->IsTerminal()) {
+      // Probably best place to check for two-fold draws consistently.
+      // Depth starts with 1 at root, so real depth is depth - 1.
+      if (node->IsTwoFoldTerminal() && depth - 1 < node->GetM()) {
+        // Check whether first repetition was before root. If yes, remove
+        // terminal status of node and revert all visits in the tree.
+        // Length of repetition was stored in m_.
+        int depth_counter = 0;
+        // Cache node's values as we reset them in the process. We could
+        // manually set wl and d, but if we want to reuse this for reverting
+        // other terminal nodes this is the way to go.
+        const auto wl = node->GetWL();
+        const auto d = node->GetD();
+        const auto m = node->GetM();
+        const auto terminal_visits = node->GetN();
+        LOGFILE << "== reverting " << terminal_visits << " visits to " <<
```

Comment:
I strongly suggest removing this log line before submission.

Reply:
Done.